Source/DafnyLanguageServer/Workspace/Notifications/VerificationDiagnosticsParams.cs

```diff
@@ -0,0 +1,664 @@
+using OmniSharp.Extensions.LanguageServer.Protocol;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using Range = OmniSharp.Extensions.LanguageServer.Protocol.Models.Range;
+
+#pragma warning disable CS8618 // Non-nullable field must contain a non-null value when exiting constructor. Consider declaring as nullable.
+
+namespace Microsoft.Dafny.LanguageServer.Workspace.Notifications {
+  /// <summary>
+  /// DTO used to communicate the current compilation status to the LSP client.
+  /// </summary>
+  [Method(DafnyRequestNames.VerificationDiagnostics, Direction.ServerToClient)]
+  public class VerificationDiagnosticsParams : IRequest, IRequest<Unit> {
+    /// <summary>
+    /// Gets the URI of the document whose verification completed.
+    /// </summary>
+    public DocumentUri Uri { get; init; }
+
+    /// <summary>
+    /// Gets the version of the document.
+    /// </summary>
+    public int? Version { get; init; }
+
+    /// <summary>
+    /// Gets the same diagnostics as displayed in the diagnostics window
+    /// </summary>
+    public Container<Diagnostic> Diagnostics { get; init; }
+
+    /// <summary>
+    /// Returns a tree of diagnostics that can be used
+    /// First-level nodes are methods, or witness subset type checks
+    /// Second-level nodes are preconditions, postconditions, body verification status
+    /// Third level nodes are assertions inside functions 
+    /// </summary>
+    public NodeDiagnostic[] PerNodeDiagnostic { get; init; }
+
+    /// <summary>
+    /// The number of lines in the document
+    /// </summary>
+    public int LinesCount { get; init; }
+
+    public int NumberOfResolutionErrors { get; init; }
+
+    /// <summary>
+    /// Returns per-line real-time diagnostic
+    /// </summary>
+    public LineVerificationStatus[] PerLineDiagnostic { get; set; } = new LineVerificationStatus[] { };
+
+    public void RecomputePerLineDiagnostics() {
+      PerLineDiagnostic = RenderPerLineDiagnostics(this, PerNodeDiagnostic, LinesCount,
+        NumberOfResolutionErrors, Diagnostics);
+    }
+
+    static LineVerificationStatus[] RenderPerLineDiagnostics(
+      VerificationDiagnosticsParams verificationDiagnosticsParams,
+      NodeDiagnostic[] perNodeDiagnostic,
+      int numberOfLines,
+      int numberOfResolutionErrors,
+      Container<Diagnostic> diagnostics
+    ) {
+      var result = new LineVerificationStatus[numberOfLines];
+
+      if (perNodeDiagnostic.Length == 0) {
+        for (var line = 0; line < numberOfLines; line++) {
+          result[line] = LineVerificationStatus.Verified;
+        }
+
+        return result;
+      }
+
+      // Render node content into lines.
+      foreach (var nodeDiagnostic in perNodeDiagnostic) {
+        if (nodeDiagnostic.Filename == verificationDiagnosticsParams.Uri.GetFileSystemPath() ||
+            "untitled:" + nodeDiagnostic.Filename == verificationDiagnosticsParams.Uri) {
+          nodeDiagnostic.RenderInto(result);
+        }
+      }
+
```

Comment:
In order to not distract users, we fill the gap between functions and methods with "verified" status if either side is verified. This can be implemented by a forward and backward pass.

Reply:
Seems like you can do this on the IDE side, by having any code that's not covered by verification diagnostics be marked as verified.